File: iosApp/iosApp/Pages/StopDetails/StopDetailsFilteredView.swift
//
//  StopDetailsFilteredView.swift
//  iosApp
//
//  Created by esimon on 11/25/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import os
import Shared
import SwiftUI

struct StopDetailsFilteredView: View {
    @ObserveInjection var inject
    var stopId: String
    var stopFilter: StopDetailsFilter
    var tripFilter: TripDetailsFilter?

    var routeData: StopDetailsViewModel.RouteData?
    var alerts: AlertsStreamDataResponse?
    var favorites: Favorites
    var global: GlobalResponse?
    var now: Date

    var onUpdateFavorites: () -> Void
    var setStopFilter: (StopDetailsFilter?) -> Void
    var setTripFilter: (TripDetailsFilter?) -> Void
    var navCallbacks: NavigationCallbacks

    var errorBannerVM: IErrorBannerViewModel
    var mapVM: IMapViewModel
    var stopDetailsVM: IStopDetailsViewModel
    var tripDetailsVM: ITripDetailsViewModel
    var favoritesVM: IFavoritesViewModel

    @State var inSaveFavoritesFlow = false
    @State var alertSummaries: [String: AlertSummary?] = [:]

    @ObservedObject var fcmTokenContainer = FcmTokenContainer.shared
    @EnvironmentObject var navManager: NavigationManager
    @EnvironmentObject var settingsCache: SettingsCache

    var analytics: Analytics = AnalyticsProvider.shared

    let inspection = Inspection<Self>()

    init(
        stopId: String,
        stopFilter: StopDetailsFilter,
        tripFilter: TripDetailsFilter?,
        routeData: StopDetailsViewModel.RouteData?,
        alerts: AlertsStreamDataResponse?,
        favorites: Favorites,
        global: GlobalResponse?,
        now: Date,
        onUpdateFavorites: @escaping () -> Void,
        setStopFilter: @escaping (StopDetailsFilter?) -> Void,
        setTripFilter: @escaping (TripDetailsFilter?) -> Void,
        navCallbacks: NavigationCallbacks,
        errorBannerVM: IErrorBannerViewModel,
        mapVM: IMapViewModel,
        stopDetailsVM: IStopDetailsViewModel,
        favoritesVM: IFavoritesViewModel = ViewModelDI().favorites,
        tripDetailsVM: ITripDetailsViewModel = ViewModelDI().tripDetails,
    ) {
        self.stopId = stopId
        self.stopFilter = stopFilter
        self.tripFilter = tripFilter
        self.routeData = routeData
        self.alerts = alerts
        self.favorites = favorites
        self.global = global
        self.now = now
        self.onUpdateFavorites = onUpdateFavorites
        self.setStopFilter = setStopFilter
        self.setTripFilter = setTripFilter
        self.navCallbacks = navCallbacks
        self.errorBannerVM = errorBannerVM
        self.mapVM = mapVM
        self.stopDetailsVM = stopDetailsVM
        self.favoritesVM = favoritesVM
        self.tripDetailsVM = tripDetailsVM
    }

    var isFavorite: Bool { if let routeStopDirection { favorites.isFavorite(routeStopDirection) } else { false }}
    var nowInstant: EasternTimeInstant { now.toEasternInstant() }
    var routeStopDirection: RouteStopDirection? { .init(
        route: stopFilter.routeId,
        stop: stopId,
        direction: stopFilter.directionId
    ) }
    var stop: Stop? { global?.getStop(stopId: stopId) }

    var lineOrRoute: LineOrRoute? { global?.getLineOrRoute(lineOrRouteId: stopFilter.routeId) }

    var realtimeStopData: RouteCardData.RouteStopData? {
        if case let .filtered(data) = onEnum(of: routeData), routeData?.filters.stopId == stopId,
           routeData?.filters.stopFilter?.routeId == stopFilter.routeId {
            data.stopData
        } else {
            nil
        }
    }

    var tripPageFilter: TripDetailsPageFilter? {
        if let tripFilter {
            .init(stopId: stopId, stopFilter: stopFilter, tripFilter: tripFilter)
        } else {
            nil
        }
    }

    var leaf: RouteCardData.Leaf? {
        realtimeStopData?.data
            .first {
                $0.stop.id == stopId && $0.lineOrRoute.id == stopFilter.routeId && $0.directionId == stopFilter
                    .directionId
            }
    }

    var body: some View {
        VStack(spacing: 0) {
            if let lineOrRoute, let global, let stop {
                loaded(lineOrRoute: lineOrRoute, stop: stop, global: global)
            } else {
                loading()
            }
        }
        .task {
            // There's no good way to know if favorites have been updated through the edit favorites flow,
            // so reload them any time this page is opened
            onUpdateFavorites()
            for await model in stopDetailsVM.models {
                alertSummaries = model.alertSummaries as? [String: AlertSummary?] ?? [:]
            }
        }
        .manageVM(
            tripDetailsVM,
            alerts: alerts,
            context: .stopDetails,
            filters: tripPageFilter,
        )
        .accessibilityHidden(inSaveFavoritesFlow)
        .onReceive(inspection.notice) { inspection.visit(self, $0) }
        .enableInjection()
    }

    @ViewBuilder
    func loaded(lineOrRoute: LineOrRoute, stop: Stop, global: GlobalResponse) -> some View {
        let allPatternsForStop: [RoutePattern] = global.getPatternsFor(stopId: stopId, lineOrRoute: lineOrRoute)

        let directions: [Direction] = lineOrRoute.directions(
            globalData: global,
            stop: stop,
            patterns: allPatternsForStop.filter { pattern in
                pattern.isTypical()
            }
        )

        let availableDirections = directions.filter {
            !stop.isLastStopForAllPatterns(directionId: $0.id, patterns: allPatternsForStop, global: global)
        }

        VStack(spacing: 0) {
            ZStack {
                Color.fill2.ignoresSafeArea(.all)
                header(
                    lineOrRoute: lineOrRoute,
                    directions: directions,
                    availableDirections: availableDirections
                )
            }
            .fixedSize(horizontal: false, vertical: true)

            ZStack(alignment: .top) {
                Color(hex: lineOrRoute.backgroundColor).ignoresSafeArea(.all)
                Rectangle()
                    .fill(Color.halo)
                    .frame(height: 2)
                    .frame(maxWidth: .infinity)

                ScrollView(.vertical, showsIndicators: false) {
                    VStack(spacing: 16) {
                        DirectionPicker(
                            availableDirections: availableDirections.map(\.id),
                            directions: directions,
                            route: lineOrRoute.sortRoute,
                            selectedDirectionId: stopFilter.directionId,
                            updateDirectionId: { setStopFilter(StopDetailsFilter(
                                routeId: stopFilter.routeId,
                                directionId: $0
                            )) }
                        )
                        .fixedSize(horizontal: false, vertical: true)
                        .padding([.horizontal, .top], 16)
                        .padding(.bottom, 6)
                        .dynamicTypeSize(...DynamicTypeSize.accessibility1)

                        departures(directions: directions)
                    }
                }
            }
        }
    }

    @ViewBuilder
    func departures(directions: [Direction]) -> some View {
        if let leaf {
            StopDetailsFilteredDepartureDetails(
                stopId: stopId,
                stopFilter: stopFilter,
                tripFilter: tripFilter,
                setStopFilter: setStopFilter,
                setTripFilter: setTripFilter,
                leaf: leaf,
                alertSummaries: alertSummaries,
                selectedDirection: directions[Int(stopFilter.directionId)],
                favorite: isFavorite,
                now: now.toEasternInstant(),
                errorBannerVM: errorBannerVM,
                mapVM: mapVM,
                stopDetailsVM: stopDetailsVM,
            )
        } else {
            loadingDepartures()
        }
    }

    @ViewBuilder
    func loading() -> some View {
        let routeData = LoadingPlaceholders.shared.routeCardData(
            routeId: stopFilter.routeId,
            trips: 10,
            context: .stopDetailsFiltered,
            now: nowInstant
        )

        return loaded(
            lineOrRoute: routeData.lineOrRoute,
            stop: routeData.stopData.first!.stop,
            global: GlobalResponse(objects: ObjectCollectionBuilder())
        ).loadingPlaceholder()
    }

    func loadingDepartures() -> some View {
        let routeData = LoadingPlaceholders.shared.routeCardData(
            routeId: stopFilter.routeId,
            trips: 10,
            context: .stopDetailsFiltered,
            now: nowInstant
        )

        return StopDetailsFilteredDepartureDetails(
            stopId: stopId,
            stopFilter: stopFilter,
            tripFilter: tripFilter,
            setStopFilter: setStopFilter,
            setTripFilter: setTripFilter,
            leaf: routeData.stopData.first!.data.first!,
            alertSummaries: alertSummaries,
            selectedDirection: routeData.stopData.first!.directions[Int(stopFilter.directionId)],
            favorite: isFavorite,
            now: now.toEasternInstant(),
            errorBannerVM: errorBannerVM,
            mapVM: mapVM,
            stopDetailsVM: stopDetailsVM,
        ).loadingPlaceholder()
    }

    @ViewBuilder
    func header(lineOrRoute: LineOrRoute, directions _: [Direction], availableDirections: [Direction]) -> some View {
        let line: Line? = switch onEnum(of: lineOrRoute) {
        case let .line(line): line.line
        default: nil
        }
        VStack(spacing: 0) {
            if let stop, let routeStopDirection, inSaveFavoritesFlow {
                SaveFavoritesFlow(
                    lineOrRoute: lineOrRoute,
                    stop: stop,
                    directions: availableDirections,
                    selectedDirection: routeStopDirection.direction,
                    context: .stopDetails,
                    global: global,
                    isFavorite: { favorites.isFavorite($0) },
                    updateFavorites: { updatedValues in
                        favoritesVM.updateFavorites(
                            updatedFavorites: updatedValues,
                            context: .stopDetails,
                            defaultDirection: routeStopDirection.direction,
                            fcmToken: fcmTokenContainer.token,
                        )
                    },
                    onClose: { inSaveFavoritesFlow = false },
                    pushNavEntry: { navManager.pushNavEntry($0) },
                )
            }

            VStack(spacing: 0) {
                StopDetailsFilteredHeader(
                    route: lineOrRoute.sortRoute,
                    line: line,
                    stop: stop,
                    direction: stopFilter.directionId,
                    isFavorite: isFavorite,
                    onFavorite: { inSaveFavoritesFlow = true },
                    navCallbacks: navCallbacks,
                )
                ErrorBanner(errorBannerVM, padding: .init([.horizontal, .bottom], 16))
                DebugView {
                    VStack(alignment: .leading) {
                        Text(verbatim: "stop id: \(stopId)")
                        Text(verbatim: "trip id: \(tripFilter?.tripId ?? "nil")")
                        Text(verbatim: "vehicle id: \(tripFilter?.vehicleId ?? "nil")")
                    }
                }
                .padding(.horizontal, 6)
            }
            .fixedSize(horizontal: false, vertical: true)
            .dynamicTypeSize(...DynamicTypeSize.accessibility1)
        }
    }
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/stopDetails/StopDetailsFilteredView.kt
package com.mbta.tid.mbta_app.android.stopDetails

import androidx.compose.foundation.ScrollState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.zIndex
import com.mbta.tid.mbta_app.android.ModalRoutes
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.component.DebugView
import com.mbta.tid.mbta_app.android.component.ErrorBanner
import com.mbta.tid.mbta_app.android.component.SaveFavoritesFlow
import com.mbta.tid.mbta_app.android.state.getGlobalData
import com.mbta.tid.mbta_app.android.util.IsLoadingSheetContents
import com.mbta.tid.mbta_app.android.util.fromHex
import com.mbta.tid.mbta_app.android.util.modifiers.loading
import com.mbta.tid.mbta_app.model.Direction
import com.mbta.tid.mbta_app.model.FavoriteSettings
import com.mbta.tid.mbta_app.model.LineOrRoute
import com.mbta.tid.mbta_app.model.LoadingPlaceholders
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder
import com.mbta.tid.mbta_app.model.RouteCardData
import com.mbta.tid.mbta_app.model.RoutePattern
import com.mbta.tid.mbta_app.model.RouteStopDirection
import com.mbta.tid.mbta_app.model.Stop
import com.mbta.tid.mbta_app.model.StopDetailsFilter
import com.mbta.tid.mbta_app.model.TripDetailsFilter
import com.mbta.tid.mbta_app.model.response.AlertsStreamDataResponse
import com.mbta.tid.mbta_app.model.response.GlobalResponse
import com.mbta.tid.mbta_app.repositories.ErrorKey
import com.mbta.tid.mbta_app.routes.SheetRoutes
import com.mbta.tid.mbta_app.usecases.EditFavoritesContext
import com.mbta.tid.mbta_app.utils.EasternTimeInstant
import com.mbta.tid.mbta_app.utils.NavigationCallbacks
import com.mbta.tid.mbta_app.viewModel.IErrorBannerViewModel
import com.mbta.tid.mbta_app.viewModel.IStopDetailsViewModel
import com.mbta.tid.mbta_app.viewModel.StopDetailsViewModel
import org.koin.compose.koinInject

@Composable
fun StopDetailsFilteredView(
    stopId: String,
    stopFilter: StopDetailsFilter,
    tripFilter: TripDetailsFilter?,
    allAlerts: AlertsStreamDataResponse?,
    now: EasternTimeInstant,
    isFavorite: (RouteStopDirection) -> Boolean?,
    updateFavorites: (Map<RouteStopDirection, FavoriteSettings?>, Int) -> Unit,
    navCallbacks: NavigationCallbacks,
    updateStopFilter: (StopDetailsFilter?) -> Unit,
    updateTripFilter: (TripDetailsFilter?) -> Unit,
    tileScrollState: ScrollState,
    openModal: (ModalRoutes) -> Unit,
    openSheetRoute: (SheetRoutes) -> Unit,
    errorBannerViewModel: IErrorBannerViewModel,
    stopDetailsViewModel: IStopDetailsViewModel = koinInject(),
) {
    val state by stopDetailsViewModel.models.collectAsState()
    val global: GlobalResponse? =
        getGlobalData(ErrorKey(setOf(SheetRoutes.StopDetails::class), "StopDetailsFilteredView"))

    val lineOrRoute: LineOrRoute? = global?.getLineOrRoute(stopFilter.routeId)

    val stop = global?.getStop(stopId)
    val realtimeStopData =
        when (val data = state.routeData) {
            is StopDetailsViewModel.RouteData.Filtered -> data.stopData
            else -> null
        }

    val leaf =
        realtimeStopData?.data?.firstOrNull {
            it.stop.id == stopId &&
                it.lineOrRoute.id == stopFilter.routeId &&
                it.direction.id == stopFilter.directionId
        }

    var inSaveFavoritesFlow by rememberSaveable { mutableStateOf(false) }

    @Composable
    fun Header(lineOrRoute: LineOrRoute, stop: Stop, availableDirections: List<Direction>) {
        val rsd = RouteStopDirection(lineOrRoute.id, stop.id, stopFilter.directionId)

        return Column(verticalArrangement = Arrangement.spacedBy(0.dp)) {
            if (inSaveFavoritesFlow) {
                SaveFavoritesFlow(
                    lineOrRoute,
                    stop,
                    availableDirections,
                    selectedDirection = stopFilter.directionId,
                    context = EditFavoritesContext.StopDetails,
                    updateFavorites = { newValues ->
                        updateFavorites(newValues, stopFilter.directionId)
                    },
                    isFavorite = { rsd -> isFavorite(rsd) ?: false },
                    onClose = { inSaveFavoritesFlow = false },
                    openModal = openModal,
                )
            }
            StopDetailsFilteredHeader(
                lineOrRoute.sortRoute,
                (lineOrRoute as? LineOrRoute.Line)?.line,
                stop,
                isFavorite = isFavorite(rsd),
                onFavorite = { inSaveFavoritesFlow = true },
                navCallbacks,
            )

            ErrorBanner(errorBannerViewModel, Modifier.padding(bottom = 16.dp))
            DebugView(
                content = {
                    Column(Modifier.align(Alignment.Start), horizontalAlignment = Alignment.Start) {
                        Text("stop id: $stopId")
                        Text("trip id: ${tripFilter?.tripId ?: "null"}")
                        Text("vehicle id: ${tripFilter?.vehicleId ?: "null"}")
                    }
                }
            )
        }
    }

    @Composable
    fun LoadingDepartures() {
        CompositionLocalProvider(IsLoadingSheetContents provides true) {
            Column(modifier = Modifier.loading()) {
                val routeData =
                    LoadingPlaceholders.routeCardData(
                        stopFilter.routeId,
                        trips = 10,
                        RouteCardData.Context.StopDetailsFiltered,
                        now,
                    )
                val stopData = routeData.stopData.single()
                val placeholderLeaf = stopData.data.first()

                StopDetailsFilteredDeparturesView(
                    stopId = stopId,
                    stopFilter = stopFilter,
                    tripFilter = tripFilter,
                    leaf = placeholderLeaf,
                    selectedDirection = stopData.data.first().direction,
                    allAlerts = AlertsStreamDataResponse(emptyMap()),
                    now = now,
                    updateTripFilter = {},
                    tileScrollState = rememberScrollState(),
                    isFavorite = false,
                    openModal = {},
                    openSheetRoute = {},
                )
            }
        }
    }

    @Composable
    fun Loaded(lineOrRoute: LineOrRoute, stop: Stop, global: GlobalResponse) {
        val routeHex: String = lineOrRoute.backgroundColor

        val routeColor: Color = Color.fromHex(routeHex)

        val allPatternsForStop: List<RoutePattern> = global.getPatternsFor(stopId, lineOrRoute)
        val directions: List<Direction> =
            lineOrRoute.directions(global, stop, allPatternsForStop.filter { it.isTypical() })

        val availableDirections = directions.filter {
            !stop.isLastStopForAllPatterns(it.id, allPatternsForStop, global)
        }

        val rsd = RouteStopDirection(lineOrRoute.id, stop.id, stopFilter.directionId)

        Column(verticalArrangement = Arrangement.spacedBy(0.dp)) {
            Header(lineOrRoute, stop, availableDirections)
            Box(Modifier.fillMaxSize().background(routeColor)) {
                HorizontalDivider(
                    Modifier.fillMaxWidth().zIndex(1f).border(2.dp, colorResource(R.color.halo))
                )
                Column(
                    Modifier.fillMaxSize()
                        .verticalScroll(rememberScrollState())
                        .padding(top = 14.dp),
                    verticalArrangement = Arrangement.spacedBy(12.dp),
                ) {
                    DirectionPicker(
                        availableDirections = availableDirections.map { it.id },
                        directions = directions,
                        route = lineOrRoute.sortRoute,
                        selectedDirectionId = stopFilter.directionId,
                        updateDirectionId = { updateStopFilter(stopFilter.copy(directionId = it)) },
                        modifier = Modifier.padding(horizontal = 10.dp),
                    )

                    if (leaf != null) {
                        StopDetailsFilteredDeparturesView(
                            stopId = stopId,
                            stopFilter = stopFilter,
                            tripFilter = tripFilter,
                            leaf = leaf,
                            selectedDirection = directions[stopFilter.directionId],
                            allAlerts = allAlerts,
                            now = now,
                            updateTripFilter = updateTripFilter,
                            tileScrollState = tileScrollState,
                            isFavorite = isFavorite(rsd) ?: false,
                            openModal = openModal,
                            openSheetRoute = openSheetRoute,
                        )
                    } else {
                        LoadingDepartures()
                    }
                }
            }
        }
    }

    return if (global != null && lineOrRoute != null && stop != null) {
        Loaded(lineOrRoute, stop, global)
    } else {
        val routeData =
            LoadingPlaceholders.routeCardData(
                stopFilter.routeId,
                10,
                RouteCardData.Context.StopDetailsFiltered,
                now,
            )

        CompositionLocalProvider(IsLoadingSheetContents provides true) {
            Column(modifier = Modifier.loading()) {
                Loaded(
                    routeData.lineOrRoute,
                    routeData.stopData.first().stop,
                    GlobalResponse(ObjectCollectionBuilder("StopDetailFilteredViewLoading")),
                )
            }
        }
    }
}
